vql/linux/audit.go

```diff
@@ -4,90 +4,688 @@ package linux
+	if debugGoRoutines {
+		val := atomic.AddInt64(&self.wgCounter, -1)
+		fmt.Printf("audit: wgcounter-: %d [%s]\n", val, description)
+	}
+	self.wg.Done()
+}
+
+// Only allow a reference if the refcount is already elevated
+func (self *AuditWatcherService) Get() bool {
+	for {
+		refcount := atomic.LoadInt64(&self.refcount)
+
+		if refcount == 0 {
+			return false
+		}
+
+		if atomic.CompareAndSwapInt64(&self.refcount, refcount, refcount + 1) {
+			break
+		}
+	}
+
+	return true
+}
+
+func (self *AuditWatcherService) Put() {
+	atomic.AddInt64(&self.refcount, -1)
+
+	// If it's the last regular reference, we'll clean up the
+	// global reference.
+	if atomic.CompareAndSwapInt64(&self.refcount, 1, 0) {
+		gService = nil
+		self.shutdown()
+	}
+}
+
+func (self *AuditWatcherService) setupWatcherService() error {
+	var err error
+
+	for _, rule := range gBannedRules {
+		watcherRule, err := parseRule(rule)
+		if err != nil {
+			self.logger.Warn("audit: Failed to parse built-in banned rule `%s': %v", rule, err)
+			return err
+		}
+
+		self.bannedRules[watcherRule.rule] = watcherRule
+	}
+
+	self.commandClient, err = libaudit.NewAuditClient(nil)
+	if err != nil {
+		return err
+	}
+
+	status, err := self.commandClient.GetStatus()
+	if err != nil {
+		self.commandClient.Close()
+		return err
+	}
 
-type streamHandler struct {
-	scope       vfilter.Scope
-	output_chan chan vfilter.Row
+	if status.Enabled == 0 {
+		self.logger.Info("audit: enabling kernel audit subsystem")
+		err = self.commandClient.SetEnabled(true, libaudit.WaitForReply)
+		if err != nil {
+			self.commandClient.Close()
+			return err
+		}
+	}
+
+	self.listenClient, err = libaudit.NewMulticastAuditClient(nil)
+	if err != nil {
+		self.commandClient.Close()
+		return err
+	}
+
+	self.reassembler, err = libaudit.NewReassembler(5, 2*time.Second, self)
+	if err != nil {
+		self.commandClient.Close()
+		self.listenClient.Close()
+		return err
+	}
+
+	self.ctx, self.cancel = context.WithCancel(context.Background())
+
+	return nil
 }
 
-func (self *streamHandler) ReassemblyComplete(msgs []*auparse.AuditMessage) {
-	self.outputMultipleMessages(msgs)
+func getAuditWatcherService(config_obj *config_proto.Config) (*AuditWatcherService, error) {
+
+	logger := logging.GetLogger(config_obj, &logging.ClientComponent)
+
+	for {
+		svc := gService
+		if svc != nil {
+			if svc.Get() {
+				return svc, nil
+			}
+
+			svc.wg.Wait()
+		}
+
+		// One at a time, please
+		gLock.Lock()
+
+		if gService != nil {
+			gLock.Unlock()
+			continue
+		}
+
+		logger.Info("audit: creating new service instance")
+
+		auditService := &AuditWatcherService{
+			rules:		map[string]*RefcountedAuditWatcherRule{},
+			bannedRules:	map[string]*AuditWatcherRule{},
+			watchers:	map[int]*AuditWatcher{},
+			logger:		logger,
+			checkerChannel:	make(chan aucoalesce.Event),
+			refcount:	1,
+		}
+
+		err := auditService.setupWatcherService()
+		if err != nil {
+			auditService = nil
```

Comment:
This is buggy, if setupWatcherService returns this function will return nil but at the end of the function you still return auditService, nil, so 'err' should be returned at the end of the function.

Reply:
Fixed in the updated PR